File: CodeDataStartFrom20250808/src/miscellaneous_problems/T34.java
package problems_set;
// public class T34 {
//     public int[] searchRange(int[] nums, int target) {
//         int n = nums.length, l = 0, r = n - 1;
//         int[] ans = new int[2];
//         while(l < r) {
//             int mid = (r - l) >> 1 + l;
//             int val = nums[mid];
//             if(val < target) {
//                 l = mid + 1;
//             }
//             else if(val > target) {
//                 r = mid - 1;
//             }
//             else {
//                 r = mid - 1;
//             }
//         }
//         ans[0] = l >= n ? -1 : nums[l] == target ? l : -1;
//         l = 0;
//         r = n - 1;
//         while(l < r) {
//             int mid = (r - l) >> 1 + l;
//             int val = nums[mid];
//             if(val < target) {
//                 l = mid + 1;
//             }
//             else if(val > target) {
//                 r = mid - 1;
//             }
//             else {
//                 l = mid + 1;
//             }
//         }
//         ans[1] = r < 0 ? -1 : nums[r] == target ? r : -1;
//         return ans;
//     }
// }

public class T34 {
    public int[] searchRange(int[] nums, int target) {
        int x = LowerBound(nums, target), n = nums.length;
        if(x == n || nums[x] != target) {
            return new int[] {-1, -1};
        }
        int y = LowerBound(nums, target + 1);
        return new int[] {x, y - 1};
    } 
    private int LowerBound(int[] nums, int target) {
        int n = nums.length, l = 0, r = n - 1;
        while(l <= r) {
            int mid = (r - l) / 2 + l;
            if(nums[mid] < target) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return l;
    }
}

File: Leetcode/DataStartFrom20241124/T34.cpp
#include <iostream>
#include <cstdio>
#include <vector>
#include <cstring>
#include <algorithm>
using namespace std;
#define Test

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> ans;
        ans.push_back(FindFirstPos(nums, target));
        ans.push_back(FindLastPos(nums, target));
        return ans;
    }
private:
    int FindFirstPos(vector<int>& num, int t){
        int mid, n = num.size(), l = 0, r = n - 1;
        while (l <= r)
        {
            mid = l + (r - l) / 2;
            if(num[mid] >= t) r = mid - 1; // 注意此处的取等和不等号的方向！！！
            else l = mid + 1;
        }
        if(l >= 0 && l < n && num[l] == t) return l;
        return -1;
    }
    int FindLastPos(vector<int>& num, int t){
        int mid, n = num.size(), l = 0, r = n - 1;
        while (l <= r)
        {
            mid = l + (r - l) / 2;
            if(num[mid] > t) r = mid - 1;
            else l = mid + 1;
        }
        if(r >= 0 && r < n && num[r] == t) return r;
        return -1;
    }
};

int main(){
    #ifdef Test
        freopen("file.in", "r", stdin);
        freopen("file.out", "w", stdout);
    #endif

    return 0;
}